runtime/server_http_request.go

```diff
@@ -101,6 +101,16 @@ func NewServerHTTPRequest(
 		logFields = append(logFields, endpoint.contextExtractor.ExtractLogFields(ctx)...)
 	}
 
+	// Overriding the api-environment and default to production
+	apiEnvironment := apiEnvironmentDefault
+	if endpoint.config != nil &&
+		endpoint.config.ContainsKey("apiEnvironmentHeader") &&
+		r.Header.Get(endpoint.config.MustGetString("apiEnvironmentHeader")) != "" {
+		apiEnvironment = r.Header.Get(endpoint.config.MustGetString("apiEnvironmentHeader"))
+	}
+	scopeTags[scopeTagsAPIEnvironment] = apiEnvironment
+	logFields = append(logFields, zap.String(apienvironmentKey, apiEnvironment))
+
 	// Overriding the environment for shadow requests
```

Comment:
The block above is the crux. We "always" emit this tag and default it to prod unless specified in request header. The header value can also be customized at internal gateways.